### PR TITLE
(3/3) Implement external secret injection for MLFlow

### PR DESCRIPTION
**Pull request by karlschriek**
This PR includes all changes include in #25 and #26.

This PR introduces the "external-secrets" Application. In addition, it rolls out all resources needed for an MLFlow setup, including an ExternalSecret resource (which will create an vanilla Secret by pulling values from the AWS Secret Manager) for RDS username and password.

In addition, the placeholder names in setup.conf have been placed within <...>, e.g. <SETUP_CONF.CLUSTER_NAME> instead of 
SETUP_CONF.CLUSTER_NAME.

The paths to the secrets (SETUP_CONF.EXTERNAL_SECRET_NAME...) in the AWS Secret Manager as well as the Role ARN to assume in order to fetch them are implemented in "setup.conf" file in the root. This same pattern can be applied for all services that need access to secret data.

I.e.:
- Create an ExternalSecret to be rolled out with the service and add placeholders for role arn and secret names
- Define the values for those placeholders in setup.conf
- Set deployments to read their values from the vanilla Secret that will be generated
- In addition, for any non-secret configuration data, create a ConfigMap and fill with placeholders. Set deployments to read parameters from that ConfigMap

### CODE REVIEW

File: distribution/mlflow/overlays/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+namespace: mlflow
+resources:
+- ../../base
```
Comment (DavidSpek) [inline distribution/mlflow/overlays/kustomization.yaml:5]: What is the purpose of this overlay?

File: distribution/kustomization.yaml [hunk 2]
```diff
@@ -35,3 +35,5 @@ resources:
 # - argocd-applications/aws-cluster-autoscaler.yaml
 # - argocd-applications/aws-load-balancer-controller.yaml
 # - argocd-applications/aws-external-dns.yaml
+- argocd-applications/external-secret.yaml
+- argocd-applications/mlflow.yaml
```
Comment (DavidSpek) [inline distribution/kustomization.yaml:39]: Missing new line

File: distribution/mlflow/base/configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: mlflow-config
+data:
+  DB_NAME: <SETUP_CONF.RDS.DB_NAME.MLFLOW>
+  RDS_HOST: <SETUP_CONF.RDS.HOST>
+  RDS_PORT: "<SETUP_CONF.RDS.PORT>"
+  ARTIFACT_S3_BUCKET: <SETUP_CONF.ARTIFACT_S3_BUCKET>
```
Comment (DavidSpek) [inline distribution/mlflow/base/configmap.yaml:9]: Missing new line

File: distribution/mlflow/base/namespace.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    control-plane: kubeflow
+    istio-injection: enabled
+  name: mlflow
```
Comment (DavidSpek) [inline distribution/mlflow/base/namespace.yaml:7]: Missing new line

**Review by DavidSpek — COMMENTED** [review]
Overall looks good. I haven't checked the functionality of the script but I trust you have. One thing that needs fixing is the `spec.source.path` parameter in all the application specs that are using kustomize due to the folder renaming. Would also be nice if the renovate config could be included again, but  I can also fix this in a follow up PR.

File: distribution/mlflow/overlays/kustomization.yaml (hunk 1 above)
Comment (karlschriek) [reply distribution/mlflow/overlays/kustomization.yaml]: Sorry, I hadn't pushed my last two commits. Should be gone now

Comment (DavidSpek) [reply distribution/mlflow/overlays/kustomization.yaml]: Now the `kustomization.yaml` in the mlflow repo seems to be missing completely.

**Comment by karlschriek** [comment]
> Overall looks good. I haven't checked the functionality of the script but I trust you have. One thing that needs fixing is the `spec.source.path` parameter in all the application specs that are using kustomize due to the folder renaming. Would also be nice if the renovate config could be included again, but I can also fix this in a follow up PR.

I did do some checks. At least, argo-cd, external-secrets and mlflow roll out. That was my mini use-case. The rest I didn't really touch aside from putting in the placeholders.

From next week I will start working with a full Kubeflow distribution.

**Comment by DavidSpek** [comment]
If you can prepend `distribution/` to `spec.source.path` in the argocd-application specs then I think this should be good for merging.

File: distribution/mlflow/overlays/kustomization.yaml (hunk 1 above)
Comment (karlschriek) [reply distribution/mlflow/overlays/kustomization.yaml]: I removed it. The mlflow folder now just contains basic specs without any overlays so there is also no need to inject kustomize inbetween.

**Comment by karlschriek** [comment]
I would have to remove `distribution/ ` from `path` everywhere then. Which is probably the cleaner way to do it anyway. I will likely only get to it on Monday though

**Comment by DavidSpek** [comment]
Sure, no problem for me.

**Comment by karlschriek** [comment]
Actually scratch that, I never commited the part where I added `distribution/ ` to path, so I will make a quick discard and get it in today. Wife and kid are anyway not home yet

**Comment by karlschriek** [comment]
Ok, so, this has clearly been a long week for me, because I am rather slow on the uptake right now. I should have actually just committed and pushed the uncommited changes (prepending distribution/ to spec.source.path in all argocd specs - that was literally the last thing I did before making PR). 

Anyway, have done it done and commit has been pushed

**Comment by DavidSpek** [comment]
@karlschriek Thanks for pushing that quickly. I think the only thing that is now still missing the the `kustomization.yaml` in the `mlflow` directory. Can you still add that quickly or should I do it in a followup PR?

**Comment by karlschriek** [comment]
It isn't needed. ArgoCD just manages all applications in the folder. If we
want to add overlays we can add a kustomization file again

**Pull request merged**